sparta/sparta/functional/RegisterDefnsJSON.hpp

```diff
@@ -0,0 +1,235 @@
+#pragma once
+
+#include <rapidjson/document.h>
+#include <rapidjson/filereadstream.h>
+#include <rapidjson/writer.h>
+#include <rapidjson/stringbuffer.h>
+
+namespace sparta
+{
+
+class RegisterDefnsFromJSON
+{
+public:
+    RegisterDefnsFromJSON(const std::vector<std::string>& register_defns_json_filenames)
+    {
+        for (const auto& filename : register_defns_json_filenames) {
+            parse(filename);
+        }
+
+        // Add a definition that indicates the end of the array
+        register_defns_.push_back(RegisterBase::DEFINITION_END);
+    }
+
+    RegisterDefnsFromJSON(const std::string& register_defns_json_filename)
+    {
+        parse(register_defns_json_filename);
+
+        // Add a definition that indicates the end of the array
+        register_defns_.push_back(RegisterBase::DEFINITION_END);
+    }
+
+    RegisterBase::Definition* getAllDefns() {
+        return register_defns_.data();
+    }
+
+private:
+    void parse(const std::string& register_defns_json_filename)
+    {
+        // Read the file into a string
+        std::ifstream ifs(register_defns_json_filename);
+        std::stringstream ss;
+        ss << ifs.rdbuf();
+        std::string json_str = ss.str();
+
+        // Parse the JSON string
+        rapidjson::Document document;
+        document.Parse(json_str.c_str());
+
+        for (auto& item : document.GetArray()) {
+            if (item.HasMember("enabled") && !item["enabled"].GetBool()) {
```

Comment:
I think this may break backwards compatibility. @furuame can you confirm? From the functional model's POV, I believe there is a difference between "this register is invalid" and "this register is valid but not enabled for this arch".

Reply:
Yes, I can confirm. This should be fine. We handle that scenario in JSON level. The `enabled` field is used to be overridden by the other arch JSON to turn that register off. The disabled register is never created.